src/qat/purr/compiler/instructions.py

```diff
@@ -348,10 +349,11 @@ def _check_filter(self, filter):
             for target in self.quantum_targets:
                 if isinstance(target, PulseChannel):
                     dt = target.physical_channel.sample_time
-                    if not np.isclose(filter.duration, dt * (self.time // dt), atol=1e-12):
+                    no_samples_from_time = math.floor(round(self.time / dt))
+                    if not np.isclose(filter.duration, dt * no_samples_from_time, atol=1e-12):
                         raise ValueError(
                             f"Filter duration '{filter.duration}' must be equal to Acquire "
-                            f"duration '{self.time}'."
+                            f"duration '{dt * no_samples_from_time}' which was rounded."
                         )
```

Comment:
similar processing as `XEngine.calculate_duration()` https://github.com/oqc-community/qat/blob/main/src/qat/purr/compiler/execution.py#L401
It's now an instance method, you could rip it off, make it a module utility function in `instructions.py` module and use it for your case:

```
calculate_duration(self)
```

Reply:
Yeah good idea, I'll also replace it in calcaulte_duration so its done it one place